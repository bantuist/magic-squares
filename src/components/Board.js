import React from 'react';
import Cells from './Cells';
import RowTotals from './RowTotals';
import ColumnTotals from './ColumnTotals';
import styled from 'styled-components';

const Container = styled.div`
  height: ${({ gridSize }) => (gridSize * 100) + 100 }px;
  width: ${({ gridSize }) => (gridSize * 100) + 200 }px;
  display: grid;
  grid-template-columns: 1fr ${({ gridSize }) => gridSize }fr 1fr;
  grid-template-rows: 3fr 1fr;
`;

const Board = ({ gridSize, grid, totals, onIsCorrectTotal, onUpdateGrid }) => {
  const { columns, rows, diagonals } = totals;
  
  return (
    <Container gridSize={gridSize}> 
      <Cells gridSize={gridSize} grid={grid} onUpdateGrid={onUpdateGrid} />
      <RowTotals rows={rows} onIsCorrectTotal={onIsCorrectTotal} />
      <ColumnTotals 
        columns={columns} 
        diagonals={diagonals} 
        onIsCorrectTotal={onIsCorrectTotal} 
      />
    </Container>
  );
};

export default Board;