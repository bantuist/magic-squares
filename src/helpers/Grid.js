
export default class Grid {
  constructor(gridSize) {
    this.gridSize = gridSize;
    this.total = this.setTotal(gridSize);
    this.grid = {};
    this.totals = { rows: {}, columns: {}, diagonals: {} };
  }
  isEven = gridSize => gridSize % 2 === 0;
  isCenter = point => Math.ceil(this.gridSize / 2) === point;

  setTotal(gridSize) {
    let total = 0;

    for(let i = 1; i <= gridSize * gridSize; i++) {
      total += i;
    };

    return total / gridSize;
  };

  setGrid(gridSize) {
    const cell = { x: gridSize + 1, y: 0};
    
    for(let i = 0; i < gridSize * gridSize; i++) {
      cell.id = i;
      cell.value = 0;
      
      // Move to next row (x--) and reset column (y = 0) at the end of each row
      if (i % gridSize === 0) {
        cell.x--;
        cell.y = 0;
      }
      
      // Move to next column (y++)
      cell.y++;
      
      // Cell is center cell if gridSize is odd 
      // and both x and y coords are equal to half the gridSize
      if (!this.isEven(gridSize) 
        && this.isCenter(cell.x) 
        && this.isCenter(cell.y)
      ) {
        cell.diagonal = 'center';
      } else if (i % (gridSize + 1) === 0) {
        cell.diagonal = '1';
      } else if (i % (gridSize - 1) === 0) {
        cell.diagonal = '2';
      } else {
        cell.diagonal = 'non-diagonal';
      }

      this.grid = { 
        ...this.grid,
        [cell.id]: {
          ...cell
        }
      };
    }

    return this.grid;
  }

  updateGrid(id, value) {
    this.grid[id].value = value;
    return this.grid;
  }

  getTotals() {
    let { grid, gridSize } = this;
    let row, column;
    let rows, columns, diagonals;

    Object.keys(grid).forEach((key, i) => {
      let cell = grid[key];
      let { x, y, value } = cell;

      rows = rows ? rows : {};
      columns = columns ? columns : {};
      row = row ? row : 0;
      column = column ? column : 0;
      column = column ? column : 0;
      diagonals = diagonals ? diagonals : [ 0, 0 ];

      // Diagonal totals based on diagonal property
      // TODO: Switches/fall-throughs
      if (cell.diagonal === 'center') {
        diagonals[0] += cell.value;
        diagonals[1] += cell.value;
      }

      if (cell.diagonal === '1') {
        diagonals[1] += cell.value;
      } else if (cell.diagonal === '2') {
        diagonals[0] += cell.value;
      }

      // Column totals based on first row
      // Set initial column value if cell index is less than gridSize
      if (i < gridSize) {
        column = value;
        columns[y] = column;
      } else {
        columns[y] += value;
      }

      // Row totals based on grid division
      // Set initial row value if cell index is evenly divisible by the gridSize
      if (i % gridSize === 0) {
        row = value;
      } else {
        row += value;
      }

      rows[x] = row;
    });
    // console.log(rows, columns, diagonals);
    return { rows, columns, diagonals };
  }
}
