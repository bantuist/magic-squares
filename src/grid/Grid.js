import { activeElement } from '../data/fixtures';
import { triangularTotal, isEven } from '../helpers';

export default class Grid {
  constructor(gridSize) {
    this.activeElement = activeElement;
    this.cells = this.setCells(gridSize);
    this.gridSize = gridSize;
    this.solution = [ 4, 9, 2, 3, 5, 7, 8, 1, 6 ];
    // this.solution = this.getSolution(gridSize);
    this.total = this.setTotal(gridSize);
    this.totals = { columns: {}, rows: {}, diagonals: {} };
  }
  isCenter = point => Math.ceil(this.gridSize / 2) === point;

  setTotal(gridSize) {
    const cellCount = gridSize * gridSize;
    return triangularTotal(cellCount, cellCount) / gridSize;
  };

  setCells(gridSize) {
    const cell = { x: gridSize + 1, y: 0};
    
    for(let i = 0; i < gridSize * gridSize; i++) {
      cell.id = i;
      cell.value = 0;
      
      // Move to next row (x--) and reset column (y = 0) at the end of each row
      if (i % gridSize === 0) {
        cell.x--;
        cell.y = 0;
      }
      
      // Move to next column (y++)
      cell.y++;
      
      // Cell is center cell if gridSize is odd 
      // and both x and y coords are equal to half the gridSize
      if (!isEven(gridSize) 
        && this.isCenter(cell.x) 
        && this.isCenter(cell.y)
      ) {
        cell.diagonal = 'center';
      } else if (i % (gridSize + 1) === 0) {
        cell.diagonal = '1';
      } else if (i % (gridSize - 1) === 0) {
        cell.diagonal = '2';
      } else {
        cell.diagonal = 'non-diagonal';
      }

      this.cells = { 
        ...this.cells,
        [cell.id]: {
          ...cell
        }
      };
    }

    return this.cells;
  }

  updateGrid(id, value) {
    this.cells[id].value = value;
    return this.cells;
  }

  getTotals() {
    let { cells, gridSize } = this;
    let row, column;
    let rows, columns, diagonals;

    Object.keys(cells).forEach((key, i) => {
      let cell = cells[key];
      let { x, y, value } = cell;

      rows = rows ? rows : {};
      columns = columns ? columns : {};
      row = row ? row : 0;
      column = column ? column : 0;
      column = column ? column : 0;
      diagonals = diagonals ? diagonals : [ 0, 0 ];

      // Diagonal totals based on diagonal property
      // TODO: Switches/fall-throughs
      if (cell.diagonal === 'center') {
        diagonals[0] += cell.value;
        diagonals[1] += cell.value;
      }

      if (cell.diagonal === '1') {
        diagonals[1] += cell.value;
      } else if (cell.diagonal === '2') {
        diagonals[0] += cell.value;
      }

      // Column totals based on first row
      // Set initial column value if cell index is less than gridSize
      if (i < gridSize) {
        column = value;
        columns[y] = column;
      } else {
        columns[y] += value;
      }

      // Row totals based on grid division
      // Set initial row value if cell index is evenly divisible by the gridSize
      if (i % gridSize === 0) {
        row = value;
      } else {
        row += value;
      }

      rows[x] = row;
    });
    // console.log(rows, columns, diagonals);
    return { rows, columns, diagonals };
  }
}
