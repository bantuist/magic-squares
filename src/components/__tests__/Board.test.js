import React from 'react';
import { shallow } from 'enzyme';
import Board from '../Board';
import Grid from '../../helpers/Grid';

const grid = new Grid(3);
const mockSetGrid = jest.fn();
const mockUpdateGrid = jest.fn();
const props = {
  activeElement: null,
  gridSize: 3, 
  total: 0,
  grid: grid.setGrid(3),
  totals: grid.getTotals(),
  setGrid: mockSetGrid,
  updateGrid: mockUpdateGrid
};

describe('Board', () => {
  const board = shallow(<Board { ...props } />);

  it('renders without crashing', () => {
    expect(board).toMatchSnapshot();
  });
});
  