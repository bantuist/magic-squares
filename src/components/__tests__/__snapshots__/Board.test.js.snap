// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Board
    activeElement={null}
    grid={
        Object {
            "0": Object {
              "diagonal": "1",
              "id": 0,
              "value": 0,
              "x": 3,
              "y": 1,
            },
            "1": Object {
              "diagonal": "non-diagonal",
              "id": 1,
              "value": 0,
              "x": 3,
              "y": 2,
            },
            "2": Object {
              "diagonal": "2",
              "id": 2,
              "value": 0,
              "x": 3,
              "y": 3,
            },
            "3": Object {
              "diagonal": "non-diagonal",
              "id": 3,
              "value": 0,
              "x": 2,
              "y": 1,
            },
            "4": Object {
              "diagonal": "center",
              "id": 4,
              "value": 0,
              "x": 2,
              "y": 2,
            },
            "5": Object {
              "diagonal": "non-diagonal",
              "id": 5,
              "value": 0,
              "x": 2,
              "y": 3,
            },
            "6": Object {
              "diagonal": "2",
              "id": 6,
              "value": 0,
              "x": 1,
              "y": 1,
            },
            "7": Object {
              "diagonal": "non-diagonal",
              "id": 7,
              "value": 0,
              "x": 1,
              "y": 2,
            },
            "8": Object {
              "diagonal": "1",
              "id": 8,
              "value": 0,
              "x": 1,
              "y": 3,
            },
          }
    }
    gridSize={3}
    setGrid={[Function]}
    total={0}
    totals={
        Object {
            "columns": Object {
              "1": 0,
              "2": 0,
              "3": 0,
            },
            "diagonals": Array [
              0,
              0,
            ],
            "rows": Object {
              "1": 0,
              "2": 0,
              "3": 0,
            },
          }
    }
    updateGrid={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Cells
          grid={
                    Object {
                              "0": Object {
                                "diagonal": "1",
                                "id": 0,
                                "value": 0,
                                "x": 3,
                                "y": 1,
                              },
                              "1": Object {
                                "diagonal": "non-diagonal",
                                "id": 1,
                                "value": 0,
                                "x": 3,
                                "y": 2,
                              },
                              "2": Object {
                                "diagonal": "2",
                                "id": 2,
                                "value": 0,
                                "x": 3,
                                "y": 3,
                              },
                              "3": Object {
                                "diagonal": "non-diagonal",
                                "id": 3,
                                "value": 0,
                                "x": 2,
                                "y": 1,
                              },
                              "4": Object {
                                "diagonal": "center",
                                "id": 4,
                                "value": 0,
                                "x": 2,
                                "y": 2,
                              },
                              "5": Object {
                                "diagonal": "non-diagonal",
                                "id": 5,
                                "value": 0,
                                "x": 2,
                                "y": 3,
                              },
                              "6": Object {
                                "diagonal": "2",
                                "id": 6,
                                "value": 0,
                                "x": 1,
                                "y": 1,
                              },
                              "7": Object {
                                "diagonal": "non-diagonal",
                                "id": 7,
                                "value": 0,
                                "x": 1,
                                "y": 2,
                              },
                              "8": Object {
                                "diagonal": "1",
                                "id": 8,
                                "value": 0,
                                "x": 1,
                                "y": 3,
                              },
                            }
          }
          gridSize={3}
          onUpdateGrid={undefined}
/>,
        <RowTotals
          onIsCorrectTotal={undefined}
          rows={
                    Object {
                              "1": 0,
                              "2": 0,
                              "3": 0,
                            }
          }
/>,
        <ColumnTotals
          columns={
                    Object {
                              "1": 0,
                              "2": 0,
                              "3": 0,
                            }
          }
          diagonals={
                    Array [
                              0,
                              0,
                            ]
          }
          onIsCorrectTotal={undefined}
/>,
      ],
      "gridSize": 3,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "grid": Object {
            "0": Object {
              "diagonal": "1",
              "id": 0,
              "value": 0,
              "x": 3,
              "y": 1,
            },
            "1": Object {
              "diagonal": "non-diagonal",
              "id": 1,
              "value": 0,
              "x": 3,
              "y": 2,
            },
            "2": Object {
              "diagonal": "2",
              "id": 2,
              "value": 0,
              "x": 3,
              "y": 3,
            },
            "3": Object {
              "diagonal": "non-diagonal",
              "id": 3,
              "value": 0,
              "x": 2,
              "y": 1,
            },
            "4": Object {
              "diagonal": "center",
              "id": 4,
              "value": 0,
              "x": 2,
              "y": 2,
            },
            "5": Object {
              "diagonal": "non-diagonal",
              "id": 5,
              "value": 0,
              "x": 2,
              "y": 3,
            },
            "6": Object {
              "diagonal": "2",
              "id": 6,
              "value": 0,
              "x": 1,
              "y": 1,
            },
            "7": Object {
              "diagonal": "non-diagonal",
              "id": 7,
              "value": 0,
              "x": 1,
              "y": 2,
            },
            "8": Object {
              "diagonal": "1",
              "id": 8,
              "value": 0,
              "x": 1,
              "y": 3,
            },
          },
          "gridSize": 3,
          "onUpdateGrid": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "function",
        "props": Object {
          "onIsCorrectTotal": undefined,
          "rows": Object {
            "1": 0,
            "2": 0,
            "3": 0,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": null,
        "nodeType": "function",
        "props": Object {
          "columns": Object {
            "1": 0,
            "2": 0,
            "3": 0,
          },
          "diagonals": Array [
            0,
            0,
          ],
          "onIsCorrectTotal": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Cells
            grid={
                        Object {
                                    "0": Object {
                                      "diagonal": "1",
                                      "id": 0,
                                      "value": 0,
                                      "x": 3,
                                      "y": 1,
                                    },
                                    "1": Object {
                                      "diagonal": "non-diagonal",
                                      "id": 1,
                                      "value": 0,
                                      "x": 3,
                                      "y": 2,
                                    },
                                    "2": Object {
                                      "diagonal": "2",
                                      "id": 2,
                                      "value": 0,
                                      "x": 3,
                                      "y": 3,
                                    },
                                    "3": Object {
                                      "diagonal": "non-diagonal",
                                      "id": 3,
                                      "value": 0,
                                      "x": 2,
                                      "y": 1,
                                    },
                                    "4": Object {
                                      "diagonal": "center",
                                      "id": 4,
                                      "value": 0,
                                      "x": 2,
                                      "y": 2,
                                    },
                                    "5": Object {
                                      "diagonal": "non-diagonal",
                                      "id": 5,
                                      "value": 0,
                                      "x": 2,
                                      "y": 3,
                                    },
                                    "6": Object {
                                      "diagonal": "2",
                                      "id": 6,
                                      "value": 0,
                                      "x": 1,
                                      "y": 1,
                                    },
                                    "7": Object {
                                      "diagonal": "non-diagonal",
                                      "id": 7,
                                      "value": 0,
                                      "x": 1,
                                      "y": 2,
                                    },
                                    "8": Object {
                                      "diagonal": "1",
                                      "id": 8,
                                      "value": 0,
                                      "x": 1,
                                      "y": 3,
                                    },
                                  }
            }
            gridSize={3}
            onUpdateGrid={undefined}
/>,
          <RowTotals
            onIsCorrectTotal={undefined}
            rows={
                        Object {
                                    "1": 0,
                                    "2": 0,
                                    "3": 0,
                                  }
            }
/>,
          <ColumnTotals
            columns={
                        Object {
                                    "1": 0,
                                    "2": 0,
                                    "3": 0,
                                  }
            }
            diagonals={
                        Array [
                                    0,
                                    0,
                                  ]
            }
            onIsCorrectTotal={undefined}
/>,
        ],
        "gridSize": 3,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "class",
          "props": Object {
            "grid": Object {
              "0": Object {
                "diagonal": "1",
                "id": 0,
                "value": 0,
                "x": 3,
                "y": 1,
              },
              "1": Object {
                "diagonal": "non-diagonal",
                "id": 1,
                "value": 0,
                "x": 3,
                "y": 2,
              },
              "2": Object {
                "diagonal": "2",
                "id": 2,
                "value": 0,
                "x": 3,
                "y": 3,
              },
              "3": Object {
                "diagonal": "non-diagonal",
                "id": 3,
                "value": 0,
                "x": 2,
                "y": 1,
              },
              "4": Object {
                "diagonal": "center",
                "id": 4,
                "value": 0,
                "x": 2,
                "y": 2,
              },
              "5": Object {
                "diagonal": "non-diagonal",
                "id": 5,
                "value": 0,
                "x": 2,
                "y": 3,
              },
              "6": Object {
                "diagonal": "2",
                "id": 6,
                "value": 0,
                "x": 1,
                "y": 1,
              },
              "7": Object {
                "diagonal": "non-diagonal",
                "id": 7,
                "value": 0,
                "x": 1,
                "y": 2,
              },
              "8": Object {
                "diagonal": "1",
                "id": 8,
                "value": 0,
                "x": 1,
                "y": 3,
              },
            },
            "gridSize": 3,
            "onUpdateGrid": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {
            "onIsCorrectTotal": undefined,
            "rows": Object {
              "1": 0,
              "2": 0,
              "3": 0,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": null,
          "nodeType": "function",
          "props": Object {
            "columns": Object {
              "1": 0,
              "2": 0,
              "3": 0,
            },
            "diagonals": Array [
              0,
              0,
            ],
            "onIsCorrectTotal": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
